### PR TITLE
[PATCH v10] Pull request for example-cls

### PR DESCRIPTION
**Pull request by kkirankk**
## linux-generic: pcap: enable packet classification

Add classifier integration to pcap pktio. With the previous
code packets received from a pcap interface were never
classified.

Signed-off-by: Kiran Kumar K <kirankumark@marvell.com>
Reviewed-by: Matias Elo <matias.elo@nokia.com>

## example: cls: add ci pass option

Added an option to check the number of received packets per CoS.
Using -C option, the minimum number of expected packets can be specified
for each CoS. On application close, the number of received packets on
each CoS will be compared to the expected values. If received packet
count is less than the expected value, the application will exit with
an error.

Signed-off-by: Kiran Kumar K <kirankumark@marvell.com>
Reviewed-by: Matias Elo <matias.elo@nokia.com>

## example: cls: add application test script

Add test for the classifier example application. The test configures one
PMR to match source IP addresses. The test traffic is read from a
capture file using pcap pktio. The number of received packets is verified
from both the target queue and default queue. Classifier example test
is now run during 'make check'.

Signed-off-by: Kiran Kumar K <kirankumark@marvell.com>
Reviewed-by: Matias Elo <matias.elo@nokia.com>

### CODE REVIEW

File: platform/linux-generic/pktio/pcap.c
Comment (matiaselo) [inline platform/linux-generic/pktio/pcap.c:291]: `cls_classify_packet()` performs full packet parse, so `packet_parse_layer()` call becomes unnecessary when classifier is enabled.

File: platform/linux-generic/pktio/pcap.c
Comment (matiaselo) [inline platform/linux-generic/pktio/pcap.c:284]: If `new_pool != pcap->pool` the packet has to copied.

Comment (matiaselo) [reply platform/linux-generic/pktio/pcap.c:284]: `PACKET_PARSE_SEG_LEN ` -> `pkt_len `

File: example/classifier/odp_classifier_run.sh
Comment (matiaselo) [inline example/classifier/odp_classifier_run.sh:4]: Please use the following format:
```
#!/bin/bash
#
# Copyright (c) 2020, Marvell
# All rights reserved.
#
# SPDX-License-Identifier:     BSD-3-Clause
#
```

File: example/classifier/Makefile.am
Comment (matiaselo) [inline example/classifier/Makefile.am:33]: Some of the tests are still failing. You can take a look at e.g.`example/l2fwd_simple/Makefile.am` for example.

File: platform/linux-generic/test/example/classifier/pktio_env
Comment (matiaselo) [inline platform/linux-generic/test/example/classifier/pktio_env:4]: Please use the common format:
```
#!/bin/sh
#
# Copyright (c) 2020, Marvell
# All rights reserved.
#
# SPDX-License-Identifier:	BSD-3-Clause
#
```

File: platform/linux-generic/test/example/classifier/pktio_env
Comment (matiaselo) [inline platform/linux-generic/test/example/classifier/pktio_env:10]: The classier example uses only one interface.

File: platform/linux-generic/test/example/classifier/pktio_env
Comment (matiaselo) [inline platform/linux-generic/test/example/classifier/pktio_env:40]: Instead of parsing the program output, which is prone to breaking, I'd prefer adding a new argument (e.g. -C,--ci-pass) for passing the minimal/exact accepted packet counts per CoS.

File: platform/linux-generic/pktio/pcap.c
Comment (matiaselo) [inline platform/linux-generic/pktio/pcap.c:296]: You can use `odp_unlikely()` here.

File: platform/linux-generic/pktio/pcap.c
Comment (matiaselo) [inline platform/linux-generic/pktio/pcap.c:291]: No need to call `odp_packet_pool ()` function, you can use `pcap->pool` directly.

File: example/classifier/Makefile.am
Comment (matiaselo) [inline example/classifier/Makefile.am:35]: Is there some reasoning for copying pktio_env and not using a symbolic link like the existing Makefile do? These steps are already quite "messy" and I'd like to keep the section as identical as possible to make maintenance easier.

E.g. example/l2fwd_simple/Makefile.am
```
# If building out-of-tree, make check will not copy the scripts and data to the
# $(builddir) assuming that all commands are run locally. However this prevents
# running tests on a remote target using LOG_COMPILER.
# So copy all script and data files explicitly here.
all-local:
	if [ "x$(srcdir)" != "x$(builddir)" ]; then \
		for f in $(EXTRA_DIST); do \
			if [ -e $(srcdir)/$$f ]; then \
				mkdir -p $(builddir)/$$(dirname $$f); \
				cp -f $(srcdir)/$$f $(builddir)/$$f; \
			fi \
		done \
	fi
	ln -f -s $(top_srcdir)/platform/$(with_platform)/test/example/l2fwd_simple/pktio_env \
		pktio_env
clean-local:
	if [ "x$(srcdir)" != "x$(builddir)" ]; then \
		for f in $(EXTRA_DIST); do \
			rm -f $(builddir)/$$f; \
		done \
	fi

```

File: example/classifier/odp_classifier_run.sh
Comment (matiaselo) [inline example/classifier/odp_classifier_run.sh:34]: Not used, can be removed.

File: platform/linux-generic/test/example/classifier/pktio_env
Comment (matiaselo) [inline platform/linux-generic/test/example/classifier/pktio_env:11]: Typos:
```
# For linux-generic the default behavior is to create one pcap interface
# which uses udp64.pcap file to inject traffic.
```

File: platform/linux-generic/test/example/classifier/pktio_env
Comment (matiaselo) [inline platform/linux-generic/test/example/classifier/pktio_env:23]: 10 seconds should be enough to pass the tests.

File: platform/linux-generic/test/example/classifier/pktio_env
Comment (matiaselo) [inline platform/linux-generic/test/example/classifier/pktio_env:24]: No used anymore.

File: platform/linux-generic/test/example/classifier/pktio_env
Comment (matiaselo) [inline platform/linux-generic/test/example/classifier/pktio_env:38]: We have a single test interface. This print doesn't provide much value, so it could be removed.

File: platform/linux-generic/test/example/classifier/pktio_env
Comment (matiaselo) [inline platform/linux-generic/test/example/classifier/pktio_env:44]: Same as above.

File: example/classifier/odp_classifier.c
Comment (matiaselo) [inline example/classifier/odp_classifier.c:124]: To make debugging easier you could also print the cos name here.

File: platform/linux-generic/pktio/pcap.c
Comment (kkirankk) [reply platform/linux-generic/pktio/pcap.c:296]: Ack

File: platform/linux-generic/pktio/pcap.c
Comment (kkirankk) [reply platform/linux-generic/pktio/pcap.c:291]: Ack

File: example/classifier/Makefile.am
Comment (kkirankk) [reply example/classifier/Makefile.am:35]: Ack, Will change to symbolic link

File: example/classifier/odp_classifier_run.sh
Comment (kkirankk) [reply example/classifier/odp_classifier_run.sh:34]: Ack

File: platform/linux-generic/test/example/classifier/pktio_env
Comment (kkirankk) [reply platform/linux-generic/test/example/classifier/pktio_env:11]: Ack

File: platform/linux-generic/test/example/classifier/pktio_env
Comment (kkirankk) [reply platform/linux-generic/test/example/classifier/pktio_env:23]: Ack

File: platform/linux-generic/test/example/classifier/pktio_env
Comment (kkirankk) [reply platform/linux-generic/test/example/classifier/pktio_env:24]: Ack

File: platform/linux-generic/test/example/classifier/pktio_env
Comment (kkirankk) [reply platform/linux-generic/test/example/classifier/pktio_env:38]: Ack

File: platform/linux-generic/test/example/classifier/pktio_env
Comment (kkirankk) [reply platform/linux-generic/test/example/classifier/pktio_env:44]: Ack

File: example/classifier/odp_classifier.c
Comment (kkirankk) [reply example/classifier/odp_classifier.c:124]: Ack

File: example/classifier/odp_classifier.c
Comment (matiaselo) [inline example/classifier/odp_classifier.c:700]: A better place for this check would be couple lines below after the `odph_thread_join(thread_tbl, num_workers);` line.

Comment (kkirankk) [reply example/classifier/odp_classifier.c:700]: Ack

File: example/classifier/odp_classifier.c
Comment (kkirankk) [reply example/classifier/odp_classifier.c:121]: Ack

Comment (matiaselo) [reply example/classifier/odp_classifier.c:121]: The comparison should be `if (strcmp(...) == 0)`

File: example/classifier/odp_classifier.c
Comment (matiaselo) [inline example/classifier/odp_classifier.c:1257]: A bit cleaned up version:
```
		"  -C, --ci_pass <dst queue:count>\n"
		"                           Minimum acceptable packet count for a CoS destination queue.\n"
		"                           If the received packet count is smaller than this value,\n"
		"                           the application will exit with an error.\n"
		"                           E.g: -C \"queue1:100\" -C \"queue2:200\" -C \"DefaultQueue:100\"\n"
```

**Comment by matiaselo** [comment]
Marvell copyright is missing from `example/classifier/odp_classifier.c`.

File: example/classifier/odp_classifier.c
Comment (kkirankk) [reply example/classifier/odp_classifier.c:1257]: Ack

**Comment by kkirankk** [comment]
@matiaselo some traffic manager test is failing. Could you please check?
FAIL: traffic_mngr/traffic_mngr_main
PASS: shmem/shmem_main
PASS: system/system_main
============================================================================
Testsuite summary for OpenDataPlane 1.25.1.0
============================================================================
# TOTAL: 36
# PASS:  35
# SKIP:  0
# XFAIL: 0
# FAIL:  1
# XPASS: 0
# ERROR: 0
============================================================================
See test/validation/api/test-suite.log
make[5]: *** [test-suite.log] Error 1
Please report to odp@lists.opendataplane.org
============================================================================
Makefile:919: recipe for target 'test-suite.log' failed
make[5]: Leaving directory '/odp/test/validation/api'
make[4]: *** [check-TESTS] Error 2
make[3]: *** [check-am] Error 2
Makefile:1025: recipe for target 'check-TESTS' failed
make[4]: Leaving directory '/odp/test/validation/api'
Makefile:1368: recipe for target 'check-am' failed
make[3]: Leaving directory '/odp/test/validation/api'
make[2]: *** [check-recursive] Error 1
Makefile:807: recipe for target 'check-recursive' failed
make[2]: Leaving directory '/odp/test/validation/api'
make[1]: *** [check-recursive] Error 1
Makefile:468: recipe for target 'check-recursive' failed
make[1]: Leaving directory '/odp/test/validation'
make: *** [check-recursive] Error 1
Makefile:563: recipe for target 'check-recursive' failed
Error: Process completed with exit code 2.

File: example/classifier/odp_classifier.c
Comment (matiaselo) [inline example/classifier/odp_classifier.c:126]: Please use ODPH_ERR macro (plus small text tweak):
```
					ODPH_ERR("Error: Cos = %s, expected packets = %" PRIu64 ", "
					       "received packets = %" PRIu64 "\n",
```

File: example/classifier/odp_classifier.c
Comment (matiaselo) [inline example/classifier/odp_classifier.c:136]: Same here:
```
			ODPH_ERR("Error: invalid Cos:%s specified for CI pass count\n",
				 args->cpass_count[i].cos_name);
```

File: example/classifier/odp_classifier.c
Comment (matiaselo) [inline example/classifier/odp_classifier.c:86]: These two variable names are a bit confusing. I'd prefer something along:
```
ci_pass_counters ci_pass_rules;
int              num_ci_pass_rules;
```

**Comment by matiaselo** [comment]
Final clean-up to the commit messages. After this you can add my reviewed-by tags to the commits.

```
linux-generic: pcap: enable packet classification

Add classifier integration to pcap pktio. With the previous
code packets received from a pcap interface were never
classified.


example: cls: add ci pass option

Added an option to check the number of received packets per CoS.
Using -C option, the minimum number of expected packets can be specified
for each CoS. On application close, the number of received packets on
each CoS will be compared to the expected values. If received packet
count is less than the expected value, the application will exit with
an error.


example: classifier: add application test script

Add test for the classifier example application. The test configures one
PMR to match source IP addresses. The test traffic is read from a capture
file using pcap pktio. The number of received packets is verified from both
the target queue and default queue. Classifier example test is now run
during 'make check'.
```

**Comment by kkirankk** [comment]
> Final clean-up to the commit messages. After this you can add my reviewed-by tags to the commits.
> 
> ```
> linux-generic: pcap: enable packet classification
> 
> Add classifier integration to pcap pktio. With the previous
> code packets received from a pcap interface were never
> classified.
> 
> 
> example: cls: add ci pass option
> 
> Added an option to check the number of received packets per CoS.
> Using -C option, the minimum number of expected packets can be specified
> for each CoS. On application close, the number of received packets on
> each CoS will be compared to the expected values. If received packet
> count is less than the expected value, the application will exit with
> an error.
> 
> 
> example: classifier: add application test script
> 
> Add test for the classifier example application. The test configures one
> PMR to match source IP addresses. The test traffic is read from a capture
> file using pcap pktio. The number of received packets is verified from both
> the target queue and default queue. Classifier example test is now run
> during 'make check'.
> ```

Will send v9 with updated commit messages and reviewd by tag.

**Comment by matiaselo** [comment]
@kkirankk Sorry, I wrote the wrong subject line to the last commit. For uniformity it should be:
`example: cls: add application test script`

**Review by matiaselo — APPROVED** [review]

**Comment by kkirankk** [comment]
> @kkirankk Sorry, I wrote the wrong subject line to the last commit. For uniformity it should be:
> `example: cls: add application test script`

Should i send v10? or is it fine?

**Comment by matiaselo** [comment]
> > @kkirankk Sorry, I wrote the wrong subject line to the last commit. For uniformity it should be:
> > `example: cls: add application test script`
> 
> @matiaselo Should i send v10? or is it fine?

Yes, please send a V10 with the subject line fix. After that I'll merge the PR after the tests have passed.

**Pull request merged**
